### PR TITLE
[Bithumb] API Changes Fixes

### PR DESCRIPTION
**Pull request by mdvx**
The Bithumb API has changes since this was last looked at
https://apidocs.bithumb.com/

### CODE REVIEW

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/BithumbAdapters.java [hunk 1]
```diff
@@ -134,15 +141,18 @@ public static LimitOrder adaptOrder(BithumbOrder order) {
         new CurrencyPair(order.getOrderCurrency(), order.getPaymentCurrency());
     final Order.OrderType orderType = adaptOrderType(order.getType());
 
+    Order.OrderStatus status = Order.OrderStatus.UNKNOWN;
+    if (order.getUnitsRemaining().compareTo(order.getUnits()) == 0) status = Order.OrderStatus.NEW;
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/BithumbAdapters.java]: use brackets

File: xchange-bithumb/src/test/resources/org/knowm/xchange/bithumb/dto/account/example-order.json [hunk 2]
```diff
@@ -1,10 +1,10 @@
 {
-  "order_id": "1546705688665840",
+  "order_id": "C01040000000000000000",
   "order_currency": "XRP",
   "order_date": "1546705688665840",
   "payment_currency": "KRW",
   "type": "ask",
-  "status": "placed",
+  "status": "0000",
```
Comment (walec51) [inline xchange-bithumb/src/test/resources/org/knowm/xchange/bithumb/dto/account/example-order.json]: is this actually something they are returning?

in the documentation I can only see:  send，pending，success，cancel

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/BithumbAdapters.java [hunk 3]
```diff
@@ -162,13 +177,14 @@ private static UserTrade adaptUserTrade(
 
     final String units = StringUtils.remove(bithumbTransaction.getUnits(), ' ');
     return new UserTrade.Builder()
+        .id(Long.toString(bithumbTransaction.getTransferDate()))
```
[the remark this answers is not in the document]
Comment (mdvx) [reply xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/BithumbAdapters.java]: ok

File: xchange-bithumb/src/test/resources/org/knowm/xchange/bithumb/dto/account/example-order.json (hunk 2 above)
Comment (mdvx) [reply xchange-bithumb/src/test/resources/org/knowm/xchange/bithumb/dto/account/example-order.json]: "status": "0000" is the top level (did the rest call work)
"order_status": "Completed" is see on orderDetails
send，pending，success，cancel i dont see in docs

it is clear that the code we had was written for an API that changed drastically, but remained on the same endpoints

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/trade/BithumbTradeResponse.java [hunk 4]
```diff
@@ -1,82 +1,21 @@
 package org.knowm.xchange.bithumb.dto.trade;
 
 import com.fasterxml.jackson.annotation.JsonProperty;
-import java.math.BigDecimal;
-import java.util.List;
 import org.knowm.xchange.bithumb.dto.BithumbResponse;
 
-public class BithumbTradeResponse extends BithumbResponse<List<BithumbTradeResponse.BithumbTrade>> {
+public class BithumbTradeResponse extends BithumbResponse {
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/trade/BithumbTradeResponse.java]: if this does not have the data field then it should not extend from `BithumbResponse`

besides it this now does not return a list of trades then there is no sense in calling it a Trade response

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/marketdata/BithumbTransactionHistoryResponse.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import org.knowm.xchange.bithumb.dto.BithumbResponse;
+
+import java.math.BigDecimal;
+import java.text.ParseException;
+import java.util.Date;
+import java.util.List;
+
+public class BithumbTransactionHistoryResponse
+    extends BithumbResponse<List<BithumbTransactionHistoryResponse.BithumbTransactionHistory>> {
+  public static final FastDateFormat TRANSACTION_DATE_FORMAT =
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/marketdata/BithumbTransactionHistoryResponse.java]: where is this used?

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/marketdata/BithumbTransactionHistoryResponse.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    public BigDecimal getTotal() {
+      return total;
+    }
+
+    public Date getTimestamp() {
+      try {
+        return TRANSACTION_DATE_FORMAT.parse(transactionDate);
+      } catch (ParseException e) {
+        return null;
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/marketdata/BithumbTransactionHistoryResponse.java]: silent errors are pure evil

use Jackson annotations to parse make it parse the date in the desired format

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbMarketDataServiceRaw.java [hunk 7]
```diff
@@ -34,10 +35,10 @@ public BithumbOrderbook getBithumbOrderBook(CurrencyPair currencyPair) throws IO
     return orderbook.getData();
   }
 
-  public List<BithumbTransactionHistory> getBithumbTrades(CurrencyPair currencyPair)
-      throws IOException {
-    final BithumbResponse<List<BithumbTransactionHistory>> transactionHistory =
-        bithumb.transactionHistory(BithumbUtils.getBaseCurrency(currencyPair));
+  public List<BithumbTransactionHistoryResponse.BithumbTransactionHistory> getBithumbTrades(
+      CurrencyPair currencyPair) throws IOException {
+    final BithumbResponse<List<BithumbTransactionHistoryResponse.BithumbTransactionHistory>>
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbMarketDataServiceRaw.java]: those inner classes make reading all of this much worse

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbTradeService.java [hunk 8]
```diff
@@ -89,12 +100,36 @@ public UserTrades getTradeHistory(TradeHistoryParams params) throws IOException
...
+        .map(oq -> (OrderQueryParamCurrencyPair) oq)
+        .flatMap(
+            oq -> {
+              try {
+                return getBithumbOrderDetail(oq.getOrderId(), oq.getCurrencyPair()).getData()
+                    .stream()
+                    .map(detail -> BithumbAdapters.adaptOrderDetail(detail, oq.getOrderId()));
+
+              } catch (IOException e) {
+                return null;
```
Comment (walec51) [inline xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/service/BithumbTradeService.java]: silent errors are pure evil

IOException  should be thrown out

File: xchange-bithumb/src/test/resources/org/knowm/xchange/bithumb/dto/account/example-order.json (hunk 2 above)
Comment (walec51) [reply xchange-bithumb/src/test/resources/org/knowm/xchange/bithumb/dto/account/example-order.json]: ok

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/BithumbAdapters.java [hunk 9]
```diff
@@ -123,60 +130,112 @@ public static Trade adaptTrade(BithumbTransactionHistory trade, CurrencyPair cur
...
             .collect(Collectors.toList());
     return new UserTrades(userTrades, Trades.TradeSortType.SortByTimestamp);
   }
 
   private static UserTrade adaptUserTrade(
-      BithumbTransaction bithumbTransaction, CurrencyPair currencyPair) {
+      BithumbUserTransactionResponse.BithumbUserTransaction bithumbTransaction,
+      CurrencyPair currencyPair) {
 
-    final String units = StringUtils.remove(bithumbTransaction.getUnits(), ' ');
```
[the remark this answers is not in the document]
Comment (mdvx) [reply xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/BithumbAdapters.java]: good point, check moved to BithumbUserTransaction()

File: xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/trade/BithumbTradeResponse.java (hunk 4 above)
Comment (mdvx) [reply xchange-bithumb/src/main/java/org/knowm/xchange/bithumb/dto/trade/BithumbTradeResponse.java]: This is only used to return the orderId of newly placed orders, so it is the REST response to place, marketBuy, marketsell.
I have removed dependency on BithumbResponse

**Pull request merged**